pkg/logicalrestore/logicalrestore.go

```diff
@@ -241,94 +204,214 @@ func (r *LogicalRestore) applyDelta(filePath string) error {
+			}
+
+			if v.FinalLSN < r.startLSN {
+				r.curLsn = dbutils.InvalidLsn
+				r.curTx = 0
+				break
+			}
+
+			r.curTx = v.XID
+			r.curLsn = v.FinalLSN
+
+			if err := r.begin(); err != nil {
+				err = fmt.Errorf("could not begin transaction: %v", err)
+			}
 		case message.Commit:
-			sql = fmt.Sprintf("COMMIT -- %d", curTx)
+			if r.curLsn != dbutils.InvalidLsn && r.curLsn != v.LSN {
+				panic("final LSN does not match begin LSN")
+			}
+
+			if err := r.commit(); err != nil {
+				err = fmt.Errorf("could not commit transaction: %v", err)
+			}
 		case message.Origin:
 			//TODO
 		case message.Truncate:
 			//TODO
 		case message.Type:
 			//TODO
-		}
+		default:
+			if r.tx == nil {
+				break
+			}
+
+			err := r.processDMLMessage(msg)
+			if err == nil {
+				break
+			}
+			if r.tx != nil {
+				if err2 := r.rollback(); err2 != nil {
+					return fmt.Errorf("could not rollback transaction: %v", err2)
+				}
+			}
 
-		log.Println(sql)
+			return fmt.Errorf("could not process message: %v", err)
+		}
 	}
 
 	return nil
 }
 
+func (r *LogicalRestore) execSQL(sql string) error {
+	if sql == "" {
+		return nil
+	}
+
+	_, err := r.tx.ExecEx(r.ctx, sql, &pgx.QueryExOptions{SimpleProtocol: true})
+
+	return err
+}
+
+func (r *LogicalRestore) processDMLMessage(msg message.Message) (err error) {
+	switch v := msg.(type) {
+	case message.Insert:
+		err = r.execSQL(v.SQL(r.relInfo))
+	case message.Update:
+		err = r.execSQL(v.SQL(r.relInfo))
+	case message.Delete:
+		err = r.execSQL(v.SQL(r.relInfo))
+	default:
+		panic("unreachable")
+	}
+
+	return
+}
+
 func (r *LogicalRestore) applyDeltas() error {
-	deltaFiles := make(deltas, 0)
-	fileList, err := ioutil.ReadDir(r.deltaDir())
+	deltaDir := path.Join(r.baseDir, utils.TableDir(r.tblOid), tablebackup.DeltasDirName)
+
+	fileList, err := ioutil.ReadDir(deltaDir)
 	if err != nil {
 		return fmt.Errorf("could not read directory: %v", err)
 	}
+
+	segmentFiles := make(segments, 0)
 	for _, v := range fileList {
-		deltaFiles = append(deltaFiles, v.Name())
+		segmentFiles = append(segmentFiles, v.Name())
 	}
 
-	sort.Sort(deltaFiles)
+	if len(segmentFiles) == 0 {
+		log.Printf("no delta files")
+		return nil
+	}
+
+	sort.Sort(segmentFiles)
 
-	for _, deltaFile := range deltaFiles {
-		if err := r.applyDelta(path.Join(r.deltaDir(), deltaFile)); err != nil {
-			return fmt.Errorf("could not apply %q delta file: %v", deltaFile, err)
+	for _, filename := range segmentFiles {
+		if err := r.applySegmentFile(path.Join(deltaDir, filename)); err != nil {
+			return fmt.Errorf("could not apply deltas from %q file: %v", filename, err)
 		}
 	}
 
 	return nil
 }
 
 func (r *LogicalRestore) checkTableStruct() error {
+	if err := r.begin(); err != nil {
+		return err
+	}
+
 	relationInfo, err := tablebackup.FetchRelationInfo(r.tx, r.NamespacedName)
 	if err != nil {
 		return fmt.Errorf("could not fetch table info: %v", err)
 	}
 
 	if !reflect.DeepEqual(relationInfo.Columns, r.relInfo.Columns) {
```

Comment:
That has a potential to break for custom-type columns, namely when the Old of the origin and destination differ, even if the type is the same.

I am not sure relation message is a good format to pass the relation data to the restore. For one, the OIDs may not match, the other issue is that we have the "isKey" attribute that we don't use.

I'd use the minimum amount of data we need to actually restore for the table info file.